# HOMEWORK 1 PART 2

### Part 0: Load Packages and Dataset

In [1]:
# Packages
import math

In [2]:
# Load Train and Test

# Load the train.txt file
with open('train.txt', 'r', encoding='utf-8') as train_file:
    train_sentences = [line.strip() for line in train_file.readlines()]

# Test if we loaded sentences properly
for i in range(3):
    print(f"{i+1}: {train_sentences[i]}")
    
print()
    

1: Under the trust 's Fit for the Future proposals , two of the hospitals would lose their A&E units with consultant - led maternity services also possibly being centralised on one site .
2: Man charged over drugs seizure
3: Borrowing from your 401 ( k ) or withdrawing from your IRA , on the other hand , can make sense in some situations , says Michael Chasnoff , a financial planner in Cincinnati : " If going back to school helps you command a higher level of income , there is a return on your investment . "



### Part 1.1: Pre-Processing

#### Pad each sentence in the training and test corpora with start and end symbols (you can use &lt;s&gt; and &lt;/s&gt;, respectively).

In [3]:
padded_train_sentences = train_sentences


# Iterate over each sentence using a range based for loop as strings are immutable, hence we have to overwrite indeces
for i in range(len(padded_train_sentences)):
    padded_train_sentences[i] = "<s> " + padded_train_sentences[i] + " </s>"



# Test if we padded our sentences properly
for i in range(3):
    print(f"{i+1}: {padded_train_sentences[i]}")
    
print()

# for i in range(3):
#     print(f"{i+1}: {padded_test_sentences[i]}")

1: <s> Under the trust 's Fit for the Future proposals , two of the hospitals would lose their A&E units with consultant - led maternity services also possibly being centralised on one site . </s>
2: <s> Man charged over drugs seizure </s>
3: <s> Borrowing from your 401 ( k ) or withdrawing from your IRA , on the other hand , can make sense in some situations , says Michael Chasnoff , a financial planner in Cincinnati : " If going back to school helps you command a higher level of income , there is a return on your investment . " </s>



### Lowercase all words in the training and test corpora. Note that the data already has been tokenized (i.e. the punctuation has been split off words).


In [4]:
lowered_padded_train_sentences = train_sentences

for i in range(len(lowered_padded_train_sentences)):
    lowered_padded_train_sentences[i] = padded_train_sentences[i].lower() 

# for i in range(len(lowered_padded_test_sentences)):
#     lowered_padded_test_sentences[i] = lowered_padded_test_sentences[i].lower() 

# Test if we lowered sentences properly
for i in range(3):
    print(f"{i+1}: {lowered_padded_train_sentences[i]}")
    
print()




1: <s> under the trust 's fit for the future proposals , two of the hospitals would lose their a&e units with consultant - led maternity services also possibly being centralised on one site . </s>
2: <s> man charged over drugs seizure </s>
3: <s> borrowing from your 401 ( k ) or withdrawing from your ira , on the other hand , can make sense in some situations , says michael chasnoff , a financial planner in cincinnati : " if going back to school helps you command a higher level of income , there is a return on your investment . " </s>



## Count Vocabulary

In [5]:
# Obtain Vocabulary | List of unique tokens  

vocabulary = set()      # Using a set reduces the complexity into 0(n + m) rather than 0(n^2) as there is no need to check for membership of the token since sets handle duplicates

for sentence in lowered_padded_train_sentences:
    tokens = sentence.split()
    
    for token in tokens:
        vocabulary.add(token)


vocabulary = list(vocabulary)

print(f"Vocabulary Obtained ...")

Vocabulary Obtained ...


In [44]:
print(vocabulary)

['sportswoman', 'stapler', 'corroborate', 'coldhearted', 'squarepants', 'spurned', 'camus', 'tpi', 'relentlessly', 'helplessly', 'athletes', '134,000', 'lenora', '4,228', 'opperman', 'caucus', 'frontrunners', '4.45pm', 'robbing', 'deepening', 'apprentice', 'cuddly', 'copedec', 'pleaser', 'bioplastics', 'titled', 'carloads', 'vyborny', 'panthar', 'cameras', 'biko', '1,195,213', 'qualifcation', 'longtime', 'unrestricted', '4p', 'http://thescotsman.scotsman.com/sport.cfm?id=1846052007', 'shelthorpe', 'altars', 'zest', 'tyranny', 'waheed', 'modest', '1971', 'terrell', 'amerencilco', 'wrigley', '4.28', 'difranza', 'pavlick', 'houma', 'tjernigan@impaccompanies.com', 'claude', 'als', 'jammu', 'espada', 'make', 'wishful', 'readmission', 'homelink', 'weld', 'receiving_arizona', 'evinced', 'larue', 'calculate', 'quashed', 'wideouts', 'shinoda', 'variously', '77.8', 'biometrics', 'cru', 'orienteering', 'cornhuskers', 'inspectors', 'groenewald', 'sofa', 'weightloss', 'zaki', 'rossoneri', 'rencontr

In [6]:
vocabulary_size = len(vocabulary)
print(f"Vocabulary size: {vocabulary_size}")

Vocabulary size: 83045


## Count Frequencies

In [7]:
# Counting frequency of all tokens 
frequency_count = {}

for sentence in lowered_padded_train_sentences:
    tokens = sentence.split()
    
    for token in tokens:
        if token in frequency_count:
            frequency_count[token] += 1
        else:
            frequency_count[token] = 1

# for token in vocabulary:
#     print(f"{str(token).ljust(20)} : {frequency_count[token]}")

print("Frequencies counted ...")

Frequencies counted ...


In [8]:
corpus_size = 0

for token in frequency_count:
    corpus_size += frequency_count[token]

print(f"There are a total of {corpus_size} tokens in the corpus.")

There are a total of 2568210 tokens in the corpus.


In [9]:
# Identify words that occured only once and change value to the uknown token
tokens_that_appeared_only_once = []

for token in frequency_count:
    if frequency_count[token] <= 1:
        tokens_that_appeared_only_once.append(token)

#print(tokens_that_appeared_only_once)
number_of_tokens_that_appeared_only_once = len(tokens_that_appeared_only_once)
print(f"Tokens that only appeared once accounted for. There are {number_of_tokens_that_appeared_only_once} that appeared only once.")

Tokens that only appeared once accounted for. There are 41307 that appeared only once.


### Convert tokens that appeared less than once to &lt;unk&gt;

In [10]:
# Tokenize sentences once before the loop
tokenized_sentences = [sentence.split() for sentence in lowered_padded_train_sentences]

# Initialize an empty list for processed sentences
processed_train_sentences = []

# Loop over each tokenized sentence
for sentence in tokenized_sentences:
    # Create a list to hold the processed tokens
    processed_sentence = []
    
    # Loop over each token in the sentence
    for token in sentence:
        # Replace token with <unk> if it appears only once
        if token in tokens_that_appeared_only_once:
            processed_sentence.append("<unk>")
        else:
            processed_sentence.append(token)
    
    # Join the processed tokens and append to the final list
    processed_train_sentences.append(" ".join(processed_sentence))

print("Tokens that appear only once have been replaced with <unk>")


Tokens that appear only once have been replaced with <unk>


In [11]:
# Verify that you replaced all the tokens that appearaed only once with the unkown token.

unk_count = 0

for sentence in processed_train_sentences:
    sentence = sentence.split()
    
    for token in sentence:

        if token == "<unk>":
            unk_count += 1


print(f"Number of tokens that appeared only once  : {number_of_tokens_that_appeared_only_once}")
print(f"Total number of <unk> tokens in the corpus: {unk_count}")

print(f"Are they equal? : {number_of_tokens_that_appeared_only_once == unk_count}")


Number of tokens that appeared only once  : 41307
Total number of <unk> tokens in the corpus: 41307
Are they equal? : True


# Part 1.2 Training the Models

In [12]:
# Counting frequency of processed tokens
processed_frequency_count = {}

for sentence in processed_train_sentences:
    tokens = sentence.split()

    for token in tokens:
        if token in processed_frequency_count:
            processed_frequency_count[token] += 1
        else:
            processed_frequency_count[token] = 1


print("Train sentence frequencies counted ...")

Train sentence frequencies counted ...


In [13]:
n_processed_train = len(processed_frequency_count)
n_corpus_size = 0

for token in processed_frequency_count:
    n_corpus_size += processed_frequency_count[token]

print(f"Vocabulary size: {n_processed_train} \nCorpus Size: {n_corpus_size}")

Vocabulary size: 41739 
Corpus Size: 2568210


### UNIGRAM MLE
P(w) = 

$$
\frac{\text{count of\ w}}{\text{total number of words in the corpus}}
$$

In [14]:
unigram_mle_probabilities = {}

for token in processed_frequency_count:
    unigram_mle_probabilities[token] =  processed_frequency_count[token] / n_corpus_size


# Sort the unigram probabilities in descending order by probability values
top_10_unigrams = sorted(unigram_mle_probabilities.items(), key=lambda x: x[1], reverse=True)[:10]

# Print the top 10 unigrams with their probabilities, formatted to 2 decimal places
for token, prob in top_10_unigrams:
    print(f"{token} : {prob:.2f}")

the : 0.05
, : 0.04
<s> : 0.04
</s> : 0.04
. : 0.03
to : 0.02
of : 0.02
and : 0.02
a : 0.02
in : 0.02


## BIGRAM MLE
$$
P(w_2 \mid w_1) = \frac{\text{count of } (w_1, w_2)}{\text{count of } w_1}
$$


In [15]:
bigram_counts = {}

for sentence in processed_train_sentences:
    words_in_sentence = sentence.split()
    
    # Iterate over sentence with index to count bigrams correctly
    for i in range(len(words_in_sentence) - 1): # Iterate through word array only once
        w1, w2 = words_in_sentence[i], words_in_sentence[i + 1]
        
        if (w1, w2) not in bigram_counts:
            bigram_counts[(w1, w2)] = 1
        else: 
            bigram_counts[(w1, w2)] += 1

print("Bigrams counted. Highest bigram frequencies: \n")
# Sort the bigrams by their counts in descending order
sorted_bigrams = sorted(bigram_counts.items(), key=lambda x: x[1], reverse=True)

# Use a for loop to print the top 10 bigrams
for i in range(10):
    bigram, count = sorted_bigrams[i]
    print(f"{bigram} : {count}")

Bigrams counted. Highest bigram frequencies: 

('.', '</s>') : 82888
('<s>', 'the') : 12233
('of', 'the') : 11832
('in', 'the') : 10640
(',', '"') : 9598
(',', 'the') : 6998
('<unk>', ',') : 6222
('<s>', '"') : 6123
(',', 'and') : 6083
('to', 'the') : 4757


In [16]:
bigram_mle_probabilities = {}

for bigram in bigram_counts:
    bigram_mle_probabilities[bigram] = bigram_counts[bigram] / processed_frequency_count[bigram[0]]


print("Bigrams probabilities computed. Highest bigram probability: \n")
# Sort the bigrams by their counts in descending order
sorted_bigram_probability = sorted(bigram_mle_probabilities.items(), key=lambda x: x[1], reverse=True)

# Use a for loop to print the top 10 bigrams
for i in range(10):
    bigram, count = sorted_bigram_probability[i]
    print(f"{bigram} : {count}")

Bigrams probabilities computed. Highest bigram probability: 

('ã"â', '...') : 1.0
('18-year', '-') : 1.0
('852', '-') : 1.0
('2520', '-') : 1.0
('esophagus', ',') : 1.0
('conjunction', 'with') : 1.0
('psychosomatic', 'medicine') : 1.0
('experimented', 'with') : 1.0
('yellen', 'said') : 1.0
('swaths', 'of') : 1.0


## BIGRAM MLE + 1 Smoothing
$$
P(w_2 \mid w_1) = \frac{\text{count of } (w_1, w_2) + 1}{\text{count of } w_1 + V}
$$


In [17]:
# Bigram +1 smoothing
bigram_mle_plus1_probabilities = {}

# Calculate smoothed bigram probabilities
for bigram in bigram_counts:
    bigram_mle_plus1_probabilities[bigram] = (bigram_counts[bigram] + 1) / (processed_frequency_count[bigram[0]] + n_corpus_size)

# Sort the bigrams by their smoothed probabilities in descending order
sorted_bigram_plus1_probability = sorted(bigram_mle_plus1_probabilities.items(), key=lambda x: x[1], reverse=True)

# Use a for loop to print the top 10 bigrams with smoothed probabilities
for i in range(10):
    bigram, prob = sorted_bigram_plus1_probability[i]
    print(f"{bigram} : {prob:.5f}")


('.', '</s>') : 0.03121
('<s>', 'the') : 0.00459
('of', 'the') : 0.00452
('in', 'the') : 0.00408
(',', '"') : 0.00358
(',', 'the') : 0.00261
('<unk>', ',') : 0.00238
('<s>', '"') : 0.00230
(',', 'and') : 0.00227
('to', 'the') : 0.00182


# 1.3 Questions

#### 1. (5 points) How many word types (unique words) are there in the training corpus? Please include the end-of-sentence padding symbol </s> and the unknown token <unk>. Do not include the start of sentence padding symbol <s>


- Vocabulary size: 83045

#### 2. (5 points) How many word tokens are there in the training corpus? Do not include the start of  sentence padding symbol <s>.

In [18]:
preprocessed_start_token_count = frequency_count['<s>']
print(f"Number of <s> tokens: {preprocessed_start_token_count}")

corpus_size_without_start = n_corpus_size - preprocessed_start_token_count 
print(f"The number of tokens without counting for <s> = {corpus_size_without_start}")

Number of <s> tokens: 100000
The number of tokens without counting for <s> = 2468210


#### 3. (10 points) What percentage of word tokens and word types in the test corpus did not occur in training (before you mapped the unknown words to <unk> in training and test data)? Please include the padding symbol </s> in your calculations. Do not include the start of sentence padding symbol <s>.


In [19]:
# Load the test.txt file
with open('test.txt', 'r', encoding='utf-8') as test_file:
    test_sentences = [line.strip() for line in test_file.readlines()]

print("Test file loaded ...")

Test file loaded ...


In [20]:
# Counting frequency of all tokens 
test_frequency_count = {}

for sentence in test_sentences:
    tokens = sentence.split()
    
    for token in tokens:
        if token in test_frequency_count:
            test_frequency_count[token] += 1
        else:
            test_frequency_count[token] = 1

# for token in vocabulary:
#     print(f"{str(token).ljust(20)} : {frequency_count[token]}")

print("Test Frequencies counted ...")

Test Frequencies counted ...


In [21]:
test_vocab_size = len(test_frequency_count)
test_corpus_size = 0

for token in test_frequency_count:
    test_corpus_size += test_frequency_count[token]

print(f"Vocabulary size: {test_vocab_size} \nTest Corpus Size: {test_corpus_size}")

Vocabulary size: 1317 
Test Corpus Size: 2669


In [22]:
# Count all the words that did not occur in training data
not_in_train_data = []

for sentence in test_sentences:
    tokens = sentence.split()

    for token in tokens:
        if token not in vocabulary:
            not_in_train_data.append(token)

In [46]:
count_not_in_train_data = len(not_in_train_data)
print(f"There are {count_not_in_train_data} tokens not in train data.")
print(f"The proportion of words not found in test data is: {count_not_in_train_data/test_corpus_size:.3f}")

There are 494 tokens not in train data.
The proportion of words not found in test data is: 0.185


#### 4. Now replace singletons in the training data with <unk> symbol and map words (in thetest corpus) not observed in training to <unk>. What percentage of bigrams (bigram types and bigram tokens) in the test corpus did not occur in training (treat <unk> as a regular token that has been observed). Please include the padding symbol </s> in your calculations. Do not include the start of sentence padding symbol <s>.


In [24]:
# Padding sentences with <s> and </s>
for i in range(len(test_sentences)):
    test_sentences[i] = "<s> " + test_sentences[i] + " </s>"

# Lowercasing the padded sentences
lowered_padded_test_sentences = [sentence.lower() for sentence in test_sentences]
print("Padded and Lowercased Test Data")

Padded and Lowercased Test Data


In [25]:
# Initialize an empty list for processed sentences
processed_test_sentences = []
tokenized_test_sentences = [sentence.split() for sentence in lowered_padded_test_sentences]

# Loop over each tokenized sentence
for sentence in tokenized_test_sentences:
    # Create a list to hold the processed tokens
    processed_sentence = []
    
    # Loop over each token in the sentence
    for token in sentence:
        # Replace token with <unk> if it does not appear in the vocab
        if token not in vocabulary:
            processed_sentence.append("<unk>")
        else:
            processed_sentence.append(token)
    
    # Join the processed tokens and append to the final list
    processed_test_sentences.append(" ".join(processed_sentence))

# Output processed test sentences
print("Tokens not in the vocabulary have been replaced with <unk>.")

Tokens not in the vocabulary have been replaced with <unk>.


In [26]:
test_bigram_counts = {}

# Count the bigrams in the processed test sentences
for sentence in processed_test_sentences:
    words_in_sentence = sentence.split()
    
    # Iterate over the sentence with an index to count bigrams correctly
    for i in range(len(words_in_sentence) - 1):
        w1, w2 = words_in_sentence[i], words_in_sentence[i + 1]
        
        if (w1, w2) not in test_bigram_counts:
            test_bigram_counts[(w1, w2)] = 1
        else: 
            test_bigram_counts[(w1, w2)] += 1

print("Bigrams counted. Highest bigram frequencies: \n")

# Sort the bigrams by their counts in descending order
sorted_test_bigrams = sorted(test_bigram_counts.items(), key=lambda x: x[1], reverse=True)

# Use a for loop to print the top 10 bigrams
for i in range(10):
    bigram, count = sorted_test_bigrams[i]
    print(f"{bigram} : {count}")


Bigrams counted. Highest bigram frequencies: 

('.', '</s>') : 85
('<s>', 'the') : 15
('in', 'the') : 15
('of', 'the') : 13
(',', 'and') : 10
(',', '"') : 10
('<unk>', ',') : 8
('for', 'the') : 7
(',', 'the') : 6
('to', 'be') : 5


In [27]:
count_bigrams_train = len(bigram_counts)
count_bigrams_test = len(test_bigram_counts)
print(f"There are {count_bigrams_train} bigrams in train data \nThere are {count_bigrams_test} bigrams in test data")

bigrams_in_test_only = []
for bigram in test_bigram_counts:
    if bigram not in bigram_counts:
        bigrams_in_test_only.append(bigram)


count_bigrams_test_only = len(bigrams_in_test_only)
proportion_of_bigrams_in_test_only = count_bigrams_test_only/count_bigrams_test

print(f"There are {count_bigrams_test_only} bigrams that appeared in test only")
print(f"The proportion of bigrams found in test only is: {proportion_of_bigrams_in_test_only:.2f}")


There are 742294 bigrams in train data 
There are 2390 bigrams in test data
There are 657 bigrams that appeared in test only
The proportion of bigrams found in test only is: 0.27


###  Compute the log probability of the following sentence under the three models (ignore capitalization and pad each sentence as described above). Please list all of the parameters required to compute the probabilities and show the complete calculation. Which of the parameters have zero values under each model? Use log base 2 in your calculations. Map words not observed in the training corpus to the <unk> token.
`I look forward to hearing your reply`

## UNIGRAM 

In [28]:
input_sentence = ["<s>", "i", "look", "forward", "to", "hearing", "your", "reply", "</s>"]
processed_input_sentence = []

# Process input sentence
for token in input_sentence:
    if token in vocabulary:
        processed_input_sentence.append(token)
    else:
        processed_input_sentence.append("<unk>")


frequency_of_tokens_in_sentence = {}

# add input sentence to map
for token in input_sentence:
    frequency_of_tokens_in_sentence[token] = 0


for token in input_sentence:
    if token in unigram_mle_probabilities:
        frequency_of_tokens_in_sentence[token] = unigram_mle_probabilities[token]

In [29]:
print(frequency_of_tokens_in_sentence)

{'<s>': 0.03893762581720342, 'i': 0.0028576323587245593, 'look': 0.000238687646259457, 'forward': 0.00018456434637354423, 'to': 0.02065563174351007, 'hearing': 8.137963795795515e-05, 'your': 0.00047387090619536564, 'reply': 5.061891356236445e-06, '</s>': 0.03893762581720342}


In [30]:
def compute_perplexity(sentence_log_probability, num_words):
    return 2 ** (-sentence_log_probability / num_words)

In [31]:
sentence_unigram = 0

for token in frequency_of_tokens_in_sentence:
    sentence_unigram += math.log2(frequency_of_tokens_in_sentence[token])


print(f"The log Probability of the sentence is: {sentence_unigram}")
num_words = 6
perplexity = compute_perplexity(sentence_unigram, num_words)
print(f"Perplexity: {perplexity}")

The log Probability of the sentence is: -90.06992741329718
Perplexity: 33033.78286426563


## BIGRAM MLE

In [32]:
## BIGRAM
sentence_bigram_counts = {}

for i in range(len(processed_input_sentence) - 1):
    w1, w2 = processed_input_sentence[i], processed_input_sentence[i + 1]
    
    if (w1, w2) in sentence_bigram_counts:
        sentence_bigram_counts[(w1, w2)] += 1
    else:
        sentence_bigram_counts[(w1, w2)] = 1


print(sentence_bigram_counts)

{('<s>', 'i'): 1, ('i', 'look'): 1, ('look', 'forward'): 1, ('forward', 'to'): 1, ('to', 'hearing'): 1, ('hearing', 'your'): 1, ('your', 'reply'): 1, ('reply', '</s>'): 1}


In [33]:
sentence_bigram_probabilities = {}

for bigram in sentence_bigram_counts:
    if bigram in bigram_mle_probabilities:
        sentence_bigram_probabilities[bigram] = bigram_mle_probabilities[bigram]
    else:
        sentence_bigram_probabilities[bigram] = 0


print(sentence_bigram_probabilities)

{('<s>', 'i'): 0.02006, ('i', 'look'): 0.0020438751873552256, ('look', 'forward'): 0.05546492659053834, ('forward', 'to'): 0.2109704641350211, ('to', 'hearing'): 0.00011310511235107827, ('hearing', 'your'): 0, ('your', 'reply'): 0, ('reply', '</s>'): 0}


In [34]:
sentence_bigram_probability = 0

for token in sentence_bigram_probabilities:
    sentence_bigram_probability += math.log2(sentence_bigram_probabilities[token])


print(f"The probability of the sentence under Bigram MLE is {sentence_bigram_probability}")

ValueError: math domain error

### BIGRAM MLE + 1 smoothing

In [142]:
sentence_bigram_plus1_probability = 0

for token in bigram_mle_plus1_probabilities:
    sentence_bigram_plus1_probability += math.log2(bigram_mle_plus1_probabilities[token])


print(f"The log probability of the sentence under Bigram MLE with smoothing is {sentence_bigram_probability}")
num_words = 6
bigram_plus1_perplexity = compute_perplexity(sentence_bigram_plus1_probability, num_words)
print(f"Perplexity: {bigram_plus1_perplexity}")

The log probability of the sentence under Bigram MLE with smoothing is -34.10122749059451


OverflowError: (34, 'Result too large')

###  7. (20 Points) Compute the perplexity of the entire test corpus under each of the models. Discuss the differences in the results you obtained.


In [152]:
# Process TEST DATA 
unigram_mle_test_probabilities = {}

for sentence in processed_test_sentences:
    tokens = sentence.split()
    for token in tokens:
        if token not in unigram_mle_probabilities:
            if token in unigram_mle_probabilities:
                unigram_mle_probabilities[token] = unigram_mle_probabilities[token]
            else:
                unigram_mle_probabilities[token] = 0
        else:
            unigram_mle_probabilities[token] = 0

print(unigram_mle_probabilities)

{'<s>': 0, 'under': 0, 'the': 0, 'trust': 9.656531202666448e-05, "'s": 0, 'fit': 4.3610140915267835e-05, 'for': 0, 'future': 0.000239077022517629, 'proposals': 3.3875734460966976e-05, ',': 0, 'two': 0, 'of': 0, 'hospitals': 3.893762581720342e-05, 'would': 0, 'lose': 0, 'their': 0, 'a&e': 1.946881290860171e-06, 'units': 5.0229537304192413e-05, 'with': 0, 'consultant': 2.5309456781182223e-05, '-': 0, 'led': 0, 'maternity': 3.893762581720342e-06, 'services': 0, 'also': 0, 'possibly': 4.3610140915267835e-05, 'being': 0, 'centralised': 1.557505032688137e-06, 'on': 0, 'one': 0, 'site': 0, '.': 0, '</s>': 0, 'man': 0, 'charged': 9.539718325214838e-05, 'over': 0, 'drugs': 0.00010707847099730941, 'seizure': 5.061891356236445e-06, 'borrowing': 9.345030196128822e-06, 'from': 0, 'your': 0.00047387090619536564, '401': 5.8406438725805135e-06, '(': 0, 'k': 1.6743179101397472e-05, ')': 0, 'or': 0, 'withdrawing': 3.893762581720342e-06, 'ira': 4.283138839892377e-06, 'other': 0, 'hand': 0.000117981006226

In [35]:
num_tokens_test = 0

for sentence in processed_test_sentences:
    tokens = sentence.split()

    for token in tokens:
        num_tokens_test += 1

print(f"There are {num_tokens_test} in the test corpus")

There are 2869 in the test corpus


## UNIGRAM PERPLEXITY

In [39]:
unigram_mle_log_prob = 0

for token in unigram_mle_probabilities:
    if token in unigram_mle_probabilities:
        unigram_mle_log_prob += math.log2(unigram_mle_probabilities[token])
    else:
        # Handle missing tokens, you can assign a small smoothing value
        unigram_mle_log_prob += math.log2(1e-10)  # Example smoothing value

unigram_mle_L = (1/num_tokens_test) * unigram_mle_log_prob
unigram_perplexity = 2 ** -(unigram_mle_L)

print(f"Perplexity of test on using the unigram is: {unigram_perplexity}")


Perplexity of test on using the unigram is: 6.342484575892709e+80


## BIGRAM MLE PERPLEXITY

In [41]:
test_bigrams_mle_log_prob = 0
for bigram in test_bigram_counts:
    if bigram in bigram_mle_probabilities:
        test_bigrams_mle_log_prob += math.log2(bigram_mle_probabilities[bigram])
    else:
        test_bigrams_mle_log_prob += math.log2(0)


test_bigrams_mle_L  = (1/num_tokens_test)
test_bigram_mle_perplexity = 2 ** -(test_bigrams_mle_L)
print(f"Perplexity of test on Bigram MLE is: {test_bigram_mle_perplexity}") 

ValueError: math domain error

## BIGRAM + 1 PERPLEXITY

In [43]:
test_bigrams_mle_plus1_log_prob = 0
for bigram in test_bigram_counts:
    if bigram in bigram_mle_plus1_probabilities:
        test_bigrams_mle_plus1_log_prob += math.log2(bigram_mle_plus1_probabilities[bigram])
    else:
        test_bigrams_mle_plus1_log_prob += math.log2(0)


test_bigrams_mle_plus1_L  = (1/num_tokens_test)
test_bigram_mle_plus1_perplexity = 2 ** -(test_bigrams_mle_plus1_L)
print(f"Perplexity of test on Bigram MLE is: {test_bigram_mle_plus1_perplexity}")

ValueError: math domain error